### PR TITLE
Implementing Orders API with permissions

### LINKED ISSUE #3700: Implement Order api with proper permissions and relationships

### PR DESCRIPTION
**Pull request by poush**
Fixes #3700 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Implements the Order API

#### Changes proposed in this pull request:
Endpoints added:

- `/orders/` GET ONLY [ Require is_admin access ]
- `/events/<int:event_id>/orders`  GET => Event Admin,   POST=>Anyone. Only admins can provide custom status. It will be to "pending" by default.

For creating an Order it is required to create 'ticket_holders' first without any order_id then add them to relationship in the POST body of Order

- `/orders/<int:order_id>/tickets` Get tickets associated with an order 

- `/orders/<identifier>/charge` [PATCH]  to post process the the payment

Currently some payment processing part is to be completed and relationships to be added. Please review the code.

### CODE REVIEW

File: app/api/helpers/payment.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        if not credentials:
+            credentials = PayPalPaymentsManager.get_credentials(order.event)
+
+        if not credentials:
+            raise Exception('PayPal credentials have not be set correctly')
+
+        if current_app.config['TESTING']:
+            return credentials['CHECKOUT_URL']
+
+        if not currency:
```
Comment (niranjan94) [inline app/api/helpers/payment.py]: This line and the next line can be combined I guess

File: app/api/orders.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    amount = fields.Float(validate=lambda n: n > 0)
+    address = fields.Str()
+    city = fields.Str()
+    state = fields.Str(db.String)
+    country = fields.Str(required=True)
+    zipcode = fields.Str()
+    completed_at = fields.DateTime(allow_none=True)
+    transaction_id = fields.Str()
+    payment_mode = fields.Str()
+    paid_via = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    brand = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    exp_month = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    id = fields.Str(dump_only=True)
+    identifier = fields.Str(dump_only=True)
+    amount = fields.Float(validate=lambda n: n > 0)
+    address = fields.Str()
+    city = fields.Str()
+    state = fields.Str(db.String)
+    country = fields.Str(required=True)
+    zipcode = fields.Str()
+    completed_at = fields.DateTime(allow_none=True)
+    transaction_id = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+    id = fields.Str(dump_only=True)
+    identifier = fields.Str(dump_only=True)
+    amount = fields.Float(validate=lambda n: n > 0)
+    address = fields.Str()
+    city = fields.Str()
+    state = fields.Str(db.String)
+    country = fields.Str(required=True)
+    zipcode = fields.Str()
+    completed_at = fields.DateTime(allow_none=True)
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    transaction_id = fields.Str()
+    payment_mode = fields.Str()
+    paid_via = fields.Str()
+    brand = fields.Str()
+    exp_month = fields.Str()
+    exp_year = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    last4 = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/api/orders.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    status = fields.Str(validate=validate.OneOf(choices=["pending", "cancelled", "confirmed", "deleted"]))
+    discount_code_id = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: Should be available in API response

File: app/api/orders.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    payment_url = fields.Str(dump_only=True)
+    stripe_token_id = fields.Str()
```
Comment (niranjan94) [inline app/api/orders.py]: `dump_only=True`. Filled by backend.

File: app/models/order.py [hunk 11]
```diff
@@ -1,8 +1,16 @@
 import datetime
 import time
+import binascii
 
 from app.models import db
 
+def get_new_order_identifier(length=8):
+    identifier = binascii.b2a_hex(os.urandom(length / 2))
```
Comment (niranjan94) [inline app/models/order.py]: Use uuid v4 for this.

File: app/api/data_layers/ChargesLayer.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
+from flask_rest_jsonapi.data_layers.base import BaseDataLayer
+
+from app.api.helpers.exceptions import UnprocessableEntity
+from app.api.helpers.ticketing import TicketingManager
+from app.models.order import Order
+
+
+class ChargesLayer(BaseDataLayer):
+    def before_create_relationship(self, json_data, relationship_field, related_id_field, view_kwargs):
```
Comment (poush) [inline app/api/data_layers/ChargesLayer.py]: Actually, the base class is AbstractClass, so I guess we have to implement every method of it.

**Comment by poush** [comment]
@niranjan94 I tried. There will be no error if we do not implement every method but it will show only PEP8 warning. If you suggest I will change it to define only required method.

**Comment by niranjan94** [comment]
@poush PEP-8 warning we could ignore for this. Please implement only the required method.

**Comment by niranjan94** [comment]
@poush I know this has been open for a long. Sorry for not reviewing this sooner. I myself was a bit confused regarding the business logic. Now that we have the dev-handbook... Could ensure this is as per the dev handbook and resolve the conflicts here ? I'll test this locally and review asap.

**Comment by poush** [comment]
@niranjan94 Sure. I will update this. :)

File: app/api/orders.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        self_view = 'v1.order_detail'
+        self_view_kwargs = {'id': '<id>'}
+        inflect = dasherize
+
+    @post_dump
+    def generate_payment_url(self, data):
+        if 'POST' in request.method or ('GET' in request.method and 'regenerate' in request.args) and 'completed' != \
+           data["status"]:
+            if data['payment_mode'] == 'stripe':
+                data['payment_url'] = 'stripe'
```
Comment (niranjan94) [inline app/api/orders.py]: Mmm ... Let this follow a specific URI Scheme ... Would be easier for clients, if there are parsers on the client side that look for URIs

Make this `stripe://payment` (similar to an application URI scheme present in android and windows for example)

File: docs/api/api_blueprint.apib [hunk 14]
```diff
@@ -7139,8 +7138,7 @@ Update a single attendee by `id`.
                     "city": "example",
                     "state": "example",
                     "country": "IN",
-                    "ticket-id": "12",
-                    "is-checked-in": "true",
+                    "is-checked_in": "true",
```
Comment (niranjan94) [inline docs/api/api_blueprint.apib]: should be `is-checked-in`

File: docs/api/api_blueprint.apib [hunk 15]
```diff
@@ -7003,16 +7003,15 @@ Create a new attendee using an event_id.
...
-                  "is-checked-in": "false",
-                  "pdf-url": "http://example.com"
+                    "firstname": "John",
+                    "lastname": "Doe",
+                    "email": "johndoe@example.com",
+                    "address": "address",
+                    "city": "example",
+                    "state": "example",
+                    "country": "IN",
+                    "is-checked_in": "false",
```
Comment (niranjan94) [inline docs/api/api_blueprint.apib]: should be `is-checked-in`

**Comment by niranjan94** [comment]
@poush please resolve conflicts here. 

@fossasia/open-event-server-devs this is an important API. So, would be better if there is another review apart from me. I have reviewed.

**Pull request merged**